### PR TITLE
Remove libc dependency

### PR DESCRIPTION
**Pull request by Bronson-Brown-deVost**
I am using this library in a WASM program.  Because of that, I needed to remove the dependency on libc in order to compile to the WASM target.  I replaced the nextafter functions in `src/boolean/helper.rs` with my native rust implementation `float_next_after`.  You can check that package and its functionality at https://gitlab.com/bronsonbdevost/next_afterf.  All tests are still passing.

Sorry that the Cargo.lock file came along with this PR (I think it should usually be in .gitignore).  Let me know if there is anything I can do to remove that.

Thank you for this great package!

### CODE REVIEW

**Comment by bluenote10** [comment]
Really good idea, I was also not happy that introducing `nextafter` implies a libc dependency.

I had looked into implementing the functionality natively in Rust, but I wasn't sure if the approach of simply to_bits => inc/dec => from_bits works in general. The [thread on users.rust-lang.org](https://users.rust-lang.org/t/analog-of-c11-stdlib-function-nextafter/28712/3) sounded like it may not be that easy, and after skimming [this lengthy discussion](https://scicomp.stackexchange.com/questions/23191/implementing-stdnextafter-should-denormals-are-zero-mode-affect-it-if-so-ho), I put the topic on todo ;). What is your opinion on that? On first glance [what glibc is doing](https://www.sourceware.org/git/gitweb.cgi?p=glibc.git;a=blob;f=math/s_nextafter.c;h=dfa5e860c40f7ae31843d13bd48d183378a4fc6b;hb=HEAD) seems to be a bit more involved?

When calling nextafter on 0, does this approach return the smallest subnormal number (5e-324) or the smallest normal number (2.2250738585072014e-308)?

**Comment by Bronson-Brown-deVost** [comment]
Yeah, I feel a bit shaky on this ground (my PhD is in Bible and Ancient Near Eastern languages, so math is not exactly my strong suit and floating points feel a bit magical to me).  I can guarantee, however, that this particular implementation provides 5e-324 as the next value higher than 0, and -5e-324 as the next value below (you can check the tests I wrote in the float_next_after library).  It also agrees with the epsilon value hard coded into Rust for the next value larger than 1.0 (or smaller than -1.0).  Also the tests I ran with large numbers looked correct too (to my untrained eye).
  
I tried several options before the current algorithm, till I hit http://www.volkerschatz.com/science/float.html, which did in fact tell me everything I "never wanted to know about floating point but will be forced to find out."  I would love to know all the edge cases (if I missed any), but it really looks like this algorithm gives the right answers.  I wish I knew an expert who could actually confirm though.

Do you have a sense for how perfect the algorithm has to be?  I did not dig too far into your code to see how vital it was that we get the next representable float as opposed to a very nearby float.

Thank you for taking the time to look at this, and no worries at all if you don't feel comfortable pulling this.  I am new to Rust (and to deep thoughts about floating point numbers), and I am happy to keep learning.

**Comment by Bronson-Brown-deVost** [comment]
By the way, the glibc code is using bitmasking and fancy stuff like that, which is perhaps super performant, but I guess is not really within the realm of Rust's "safety" concerns.  Or at least, it was not the type of code I was interested in attempting (or able to undertake) in Rust.  Checkout Go's implementation https://golang.org/src/math/nextafter.go, which is much simpler, but maybe not always correct?

**Comment by bluenote10** [comment]
> no worries at all if you don't feel comfortable pulling this

Not at all, I'm quite confident that we can make it work, but I actually don't have maintenance rights on the repo anyway ;).

> Do you have a sense for how perfect the algorithm has to be?

I think there are only a few requirements, and even the question whether it should iterate through subnormal numbers is debatable. I've been playing around with the implementation, and so far it pretty much does exactly what is best for the algorithm.

I have forked your project and I'm extending the tests a little bit, and I have a few ideas of minor simplifications / optimizations. So far I have only found one edge case where it gives a wrong result, but I think I can fix it easily. I'll open a PR later once I have figured out how GitLab works.

**Comment by Bronson-Brown-deVost** [comment]
That's great and thanks for the help!  If we do want to avoid subnormals, maybe we can add a little check for the cases that would result in a subnormal close to zero and then just return std::f64::MIN_POSITIVE (positive or negative).

**Comment by bluenote10** [comment]
Exactly, the question of handling subnormals can still be decided within rust-geo-booleanop. Concerning next-after in general the expected behavior definitely is to iterate through subnormals, like it does now. I have prepared a PR [here](https://gitlab.com/bronsonbdevost/next_afterf/-/merge_requests/1#diff-content-b24749917179fb5e3e613ed2a703fcdcc6cdf9da).

**Comment by Bronson-Brown-deVost** [comment]
So @bluenote10 did a lot of work on the library https://crates.io/crates/float_next_after, and I think we have a very nice native solution with that crate.  If you need to avoid subnormals for some reason, we have a simple work around in mind (otherwise you can just use the crate as-is).

**Comment by bluenote10** [comment]
Yes after looking at more edge cases I also feed confident that there is no need for relying on libc any longer. Also, avoiding subnormals would actually dangerous from a correctness point of view, and would only be warranted as an optimization. What still could be done:

- Bump next-after-float to fixed version with non-reference interface.
- I just realized that the package is using underscores, but it looks like using dashes has become the standard package naming convention in Rust (and underscores only for the crates). Not sure how difficult it is though to rename a package?
- I think you are right about not committing the `Cargo.lock` for library packages, so it may be removed with this PR as well.

By the way, you may have to ping AT untoldwild eventually to get this merged (I think he does not receive automatic notifications for PRs).

**Comment by Bronson-Brown-deVost** [comment]
Many thanks, I bumped this PR to use the non-reference interface to float_next_after.  I tried renaming that crate, but to no avail; online sources suggest it is not possible, but also that underscore/hyphen doesn't matter.  I would have preferred the hyphens now, oh well. I inserted Cargo.lock into `.gitignore` as well. 

@bluenote10, would you like me to add you as an author to float_next_after?  If so, let me know the credentials you prefer.

@untoldwind, could you take a look at this PR?  It looks like everything is working well, but perhaps you have some other tests or edge cases you would like to try out.

**Comment by bluenote10** [comment]
> would you like me to add you as an author to float_next_after

Not important to me ;) only if you don't want to maintain it alone.

File: lib/Cargo.toml [hunk 1]
```diff
@@ -13,7 +13,7 @@ keywords = ["gis", "geo", "geography", "geospatial"]
 geo-types = { version = "0.4", default-features = false }
 num-traits = "0.2"
 robust = "0.1"
-float_next_after = "0.1.2"
+float_next_after = "0.1"
```
Comment (bluenote10) [inline lib/Cargo.toml:16]: Interesting, the newer version has a lower number. I though cargo makes sure that versions cannot be overwritten and can only "move forward".

Comment (Bronson-Brown-deVost) [reply lib/Cargo.toml:16]: Yeah, I changed it from 0.1.2 to 0.1 since cargo believes in semantic versioning.  If you put 0.1.2, it will use 0.1.2, but if you use 0.1 it will pull the latest version in 0.1.x (which is currently 0.1.5).  

Probably I should have changed the version to 0.2 when we changed away from the reference interface.

**Comment by untoldwind** [comment]
Thanks for that, it's really hard to keep track of all the different alternatives in the rust library sphere
